src/main/kotlin/com/sparetimedevs/suspendmongo/resilience/Resilience.kt

```diff
@@ -0,0 +1,28 @@
+package com.sparetimedevs.suspendmongo.resilience
+
+import io.github.resilience4j.bulkhead.Bulkhead
+import io.github.resilience4j.bulkhead.BulkheadConfig
+import io.github.resilience4j.retry.IntervalFunction
+import io.github.resilience4j.retry.Retry
+import io.github.resilience4j.retry.RetryConfig
+import java.util.function.Supplier
+
+class Resilience(maxConcurrentCalls: Int = 15, maxWaitMillis: Long = 3000, maxRetries: Int = 3) {
+
+    private val retryConfig = RetryConfig.custom()
+            .maxAttempts(maxRetries)
+            .intervalFunction(IntervalFunction.ofExponentialBackoff(125L, 3.0))
+            .retryOnException { true }
+            .build()
+
+    private val bulkheadConfig = BulkheadConfig.custom()
+            .maxConcurrentCalls(maxConcurrentCalls)
+            .maxWaitTime(maxWaitMillis)
+            .build()
+
+    private val retry = Retry.of("mongo-retry", retryConfig)
+    private val bulkhead = Bulkhead.of("mongo-bulkhead", bulkheadConfig)
+
+    fun <T> executeSupplier(supplier: Supplier<T?>): T? = bulkhead.executeSupplier { retry.executeSupplier(supplier) }
```

Comment:
Does this need to be Supplier<nullable type> and have a nullable return type? It leads to more questionmarks.

Reply:
Removed the nullable type from the supplier. Seems like it's happy. Removed ? from the callers.